### ISSUE TITLE
Feature importance calculate on validation set

### ISSUE THREAD

**Issue by penolove** [Microsoft/LightGBM#1262]
seems that feature importance calculates on the training set, can it apply to validation set?

**Comment by Zelazny7** [comment]
You can use the column sums of the absolute values of the predictor contributions to achieve this. When predicting, pass in `pred_contrib=true` or `predcontrib=TRUE` (R version) which will return a matrix of predictor contributions. The last column is the model intercept and can be ignored. By summing the absolute values of these contributions you can calculate importance on any data set.

**Comment by Zelazny7** [comment]
Yes, you're understanding is the same as mine! You will be calculating the Shap scores for each feature. The have the nice property of being consistent and adding up to the predicted score so I agree with you about them being better as well. However, they may take a while to compute on large datasets. I haven't done any timing tests and I'm not sure what the complexity of the operation is.

**Comment by penolove** [comment]
Thx! Is the shap feature importance mention in this paper? https://arxiv.org/abs/1706.06060

**Comment by Zelazny7** [comment]
Yes, that is what's being calculated using `pred_contrib = true` or `predcontrib=TRUE`. I had nothing to do with that paper and only exposed what the author of that paper added to LightGBM in the R package

**Comment by Zelazny7** [comment]
@Laurae2 given how the algorithm was implemented, that doesn't seem unreasonable to me. Thank you for testing the performance.
